src/satosa/micro_services/ldap_attribute_store.py

```diff
@@ -255,184 +265,199 @@ def _ldap_connection_factory(self, config):
+            )
+            msg = "Successfully connected to LDAP server"
+            satosa_logging(logger, logging.DEBUG, msg, None)
 
         except LDAPException as e:
-            msg = "Caught exception when connecting to LDAP server: {}".format(e)
+            msg = "Caught exception when connecting to LDAP server: {}"
+            msg = msg.format(e)
             satosa_logging(logger, logging.ERROR, msg, None)
             raise LdapAttributeStoreError(msg)
 
-        satosa_logging(logger, logging.DEBUG, "Successfully connected to LDAP server", None)
+        msg = "Successfully connected to LDAP server"
+        satosa_logging(logger, logging.DEBUG, msg, None)
 
         return connection
 
-    def _populate_attributes(self, config, record, context, data):
+    def _populate_attributes(self, config, record):
         """
         Use a record found in LDAP to populate attributes.
         """
-        search_return_attributes = config['search_return_attributes']
-        for attr in search_return_attributes.keys():
-            if attr in record["attributes"]:
-                if record["attributes"][attr]:
-                    data.attributes[search_return_attributes[attr]] = record["attributes"][attr]
-                    satosa_logging(
-                        logger,
-                        logging.DEBUG,
-                        "Setting internal attribute {} with values {}".format(
-                            search_return_attributes[attr],
-                            record["attributes"][attr]
-                            ),
-                        context.state
-                        )
-                else:
-                    satosa_logging(
-                        logger,
-                        logging.DEBUG,
-                        "Not setting internal attribute {} because value {} is null or empty".format(
-                            search_return_attributes[attr],
-                            record["attributes"][attr]
-                            ),
-                        context.state
-                        )
-
-    def _populate_input_for_name_id(self, config, record, context, data):
+
+        ldap_attributes = record.get("attributes", None)
+        if not ldap_attributes:
+            msg = "No attributes returned with LDAP record"
+            satosa_logging(logger, logging.DEBUG, msg, None)
+            return
+
+        ldap_to_internal_map = (
+            config["ldap_to_internal_map"]
+            if config["ldap_to_internal_map"]
+            # Deprecated configuration. Will be removed in future.
+            else config["search_return_attributes"]
+        )
+
+        attributes = defaultdict(list)
+
+        for attr, values in ldap_attributes.items():
+            internal_attr = ldap_to_internal_map.get(attr, None)
+            if not internal_attr and ';' in attr:
+                internal_attr = ldap_to_internal_map.get(attr.split(';')[0],
```

Comment:
This is strange. 

What is the format that LDAP uses to represent the attributes it returns? 
Does this mean that there are attribute-names that are concatenated by `;`?

I do not think this is the place to handle this. This should be handle in the place where `ldap_attributes` are defined. We shouldn't have to take this into account when traversing the attributes.

Reply:
The logic is necessary to properly handle LDAP attribute options. If you are not aware of LDAP attribute options see RFC 4512 and 4520, though you may find it more efficient to examine the explanation for how LDAP attribute options are leveraged in the voPerson schema at https://github.com/voperson/voperson/blob/master/voPerson.md#voperson-attribute-options.

The Python ldap3 module properly handles LDAP attribute options and the code as written makes it possible for the person deploying and configuring the microservice to map the attribute values to specific SATOSA internal attributes both when the deployer knows precisely which LDAP attribute option is use and when the deployer does not know which attribute options are in use and wants to collect values for all.

Reply:
Alright, so this means that attributes could look like:
```
{
    "some_attribute_name": "some_attribute_value",
    "uid;app-wiki": "plee@myvo.org",
    "mobile;internal": "+1 646 555 1212",
}
```

so, the `;<option>` part is optional and we ignore it for this particular processing.

What I suggest is to move this where `ldap_attributes` are defined. So,

```python
        ldap_attributes = {
            name: value
            for name_options, value in record.get("attributes", {}).items()
            for name in [name_options.split(";", 2)[0]]
        }
```

then use the `attr` from `ldap_attributes` without this check.

Reply:
Here is a use case that the code needs to cover. The Python ldap3 module returns

{
    voPersonApplicationUID;app-vidpeptid: [1128993321],
    voPersonApplicationUID;app-vidpeppn: [scottk@messproject.org],
    voPersonAffiliation;role-111: [staff],
    voPersonAffiliation: [member]
}

A deployer may need this mapped to internal attributes like this:

{
  eptid: [1128993321],
  eppn: [scottk@messproject.org],
  affiliation: [staff, member]
}
 
I don't think your suggestion covers that use case. Does it?

Reply:
Hold in mind that an attribute description consists of an attribute type and zero or more attribute options. 
I don't think there is an upper limit as to how many options you can have.

Reply:
Right/correct/agreed.

The current logic in the code accommodates that by iterating through the values and attempting to find mappings to internal SATOSA attributes, going from fine-grained mappings to less fine-grained. So if the deployer knows that a particular option is going to exist, she can make sure it gets mapped to a specific internal attribute. If the deployer doesn't know which options might be returned (or none at all), she can have all the values mapped to an internal attribute.

Reply:
So, to make this clear (at least for me), the point here is that the mapping to internal attributes could potentially include the options. As such, the match should happen when looking up the attributes - we cannot avoid this.